docs/extending/developing-code/extending/case-studies/KroneckerProduct/completing.rst

```diff
@@ -0,0 +1,141 @@
+``Completing KroneckerProduct``
+===============================
+
+.. contents::
+   :depth: 1
+   :local:
+
+
+``KroneckerProduct[]`` skeletal code
+------------------------------------
+
+Now we fill in the parts of ``KroneckerProduct`` other than the class
+definition line and its docstring.
+
+None of the other built-in Mathics Functions serve as good model to
+copy from, so we'll use a combination ``Transpose`` and
+``PauliMatrix``. The former function is selected because it takes a
+Matrix parameter, and PauliMatrix function because it is a SymPy
+function which was recently created at the time of this writing.
+
+Here is the first cut for ``KroneckerProduct``:
+
+.. code-block::
+
+    class KroneckerProduct(SympyFunction):
+        """
+	... as before
+        """
+
+        attributes = A_PROTECTED | A_READ_PROTECTED
+        summary_text = "Kronecker product"
+        sympy_name = "physics.quantum.TensorProduct"
+
+        def apply(self, mi: ListExpression, evaluation: Evaluation):
+            "KroneckerProduct[mi__]"
+            sympy_mi = [to_sympy_matrix(m) for m in mi.elements]
```

Comment:
In WMA, `KroneckerProduct` accepts a sequence (not a list) of matrices. So, the rule should be
```
    "KroneckerProduct[mi__List]"
```
and then, 
```
sympy_mi = [to_sympy_matrix(m) for m in mi.elements()]
```
Also, instead of   ``"KroneckerProduct[mi__List]"`` we can use a simpler rule
```
"KroneckerProduct[mi__]"
```
and then, inside the code, do something like
```
if not all(m.head == SymbolList for m in mi.elements):
    return None
```

Reply:
I have a hard time understanding the above. Is it correct that in version 1 it should be `mi__List` instead of `mi__`. 

And then in version 2 you suggest, dropping the `List` part take anything  add a test for List-ness (where  the `==` test should be `is`)? 

Ok, but this needs to be explained better.

Reply:
> Ok - thanks for the more detailed explanation. It is kind of what I suspected, but wasn't totally sure before.
> 
> Given this, I think what I want to do is start off with the simple-minded implementation and then add another iteration/section where this kind of thing is revised with the reasoning behind that.

Indeed, I think that this is the best for a beginner's tutorial. For `KroneckerProduct` I guess that both choices have almost the same performance in typical use cases, just because of how the complexity scales in the pattern matching (linear in the number of elements in the sequence) and exponential in the evaluation.


> The thing we should keep in mind is that even though this kind of thing is done faster as Python coding that might not always be the case in the future. Or there may come a time when the two are roughly equivalent.
> 

I have the feeling that we can improve a lot the performance of the pattern matching mechanism. However, I think that a complicated Pattern always is going to be more costly to check than a condition inside the body of the function.

> I don't see a reason in the future why `x__` couldn't be turned into an `all(...)`

Maybe that is something we can use when we implement the `DispatchRules` compilation.